Python/tier2_redundancy_eliminator_bytecodes.c

```diff
@@ -0,0 +1,239 @@
+#include "Python.h"
+#include "pycore_uops.h"
+#include "pycore_uop_ids.h"
+
+#define op(name, ...) /* NAME is ignored */
+
+typedef struct _Py_UOpsSymType _Py_UOpsSymType;
+typedef struct _Py_UOpsAbstractInterpContext _Py_UOpsAbstractInterpContext;
+typedef struct _Py_UOpsAbstractFrame _Py_UOpsAbstractFrame;
+
+static int
+dummy_func(void) {
+
+    PyCodeObject *code;
+    int oparg;
+    _Py_UOpsSymType *flag;
+    _Py_UOpsSymType *left;
+    _Py_UOpsSymType *right;
+    _Py_UOpsSymType *value;
+    _Py_UOpsSymType *res;
+    _Py_UOpsSymType *iter;
+    _Py_UOpsSymType *top;
+    _Py_UOpsSymType *bottom;
+    _Py_UOpsAbstractFrame *frame;
+    _Py_UOpsAbstractInterpContext *ctx;
+    _PyUOpInstruction *inst;
+    _PyBloomFilter *dependencies;
+    int modified;
+
+// BEGIN BYTECODES //
+
+    op(_LOAD_FAST_CHECK, (-- value)) {
+        value = GETLOCAL(oparg);
+        // We guarantee this will error - just bail and don't optimize it.
+        if (sym_is_type(value, NULL_TYPE)) {
+            goto error;
+        }
+    }
+
+    op(_LOAD_FAST, (-- value)) {
+        value = GETLOCAL(oparg);
+    }
+
+    op(_LOAD_FAST_AND_CLEAR, (-- value)) {
+        value = GETLOCAL(oparg);
+        _Py_UOpsSymType *temp = sym_init_null(ctx);
+        if (temp == NULL) {
+            goto error;
+        }
+        GETLOCAL(oparg) = temp;
+    }
+
+    op(_STORE_FAST, (value --)) {
+        GETLOCAL(oparg) = value;
+    }
+
+    op(_STORE_FAST_MAYBE_NULL, (value --)) {
+        GETLOCAL(oparg) = value;
+    }
+
+    op(_PUSH_NULL, (-- res)) {
+        res = sym_init_null(ctx);
+        if (res == NULL) {
+            goto error;
+        };
+    }
+
+    op(_GUARD_BOTH_INT, (left, right -- left:  &PYLONG_TYPE, right:  &PYLONG_TYPE)) {
+        if (sym_matches_type(left, PYLONG_TYPE, 0) &&
+            sym_matches_type(right, PYLONG_TYPE, 0)) {
+            REPLACE_OP(_NOP, 0, 0);
+        }
```

Comment:
This should promote `left` and `right` to `PYLONG_TYPE` as well.

Reply:
The type annotation automatically handles setting the type for us.

Reply:
Please make it explicit. It makes the code difficult to understand if things are added implicitly.